docs/content/sips/000-spin-test.md

```diff
@@ -0,0 +1,190 @@
+- The test itself also runs in WebAssembly allowing tests to be written in any language with WebAssembly component guest support.
+- Allows the user to modify the testing environment in a programmatic way in any language they want to use.
+- Have the test environment match functionality as much as is possible with other Spin runtimes such as found in Spin CLI, Fermyon Cloud, and SpinKube.
+
+## Proposal
+
+To achieve the goals above, we propose a new tool tentatively called `spin-test`.
+
+The `spin-test` binary would be callable from any Spin project in much the same way that `spin up` is project aware (i.e., it doesn’t need to be told where the `spin.toml` manifest is). 
+
+Invoking `spin-test` will automatically resolve where the Spin app component lives. The tool will automatically find tests that should be run and will run those tests against the Spin app component reporting whether the tests passed or failed. `spin-test` will come with basic ways of filtering tests so that a subset of tests can be run.
+
+**Note:** See the *Open Questions* section for discussion on how `spin-test` discovers tests to run.
+
+### Implementation
+
+At the core of `spin-test` is component composition. The user’s Spin app is composed together with other components to ultimately produce a component (referred to as “the composition” below) that has none of the original Spin or WASI imports. The composition has the following components internal to it:
+
+- A single component from the Spin app (see the *Open Questions* section for why only one component from a Spin app can be supported)
+- A Spin virtualization component that virtualizes all `fermyon:spin/platform` interfaces.
+- A WASI virtualization component that virtualizes all the `wasi:cli/imports` interfaces.
+- A router component that decides from a `spin.toml` manifest file whether an `incoming-request` should be routed to the Spin component or not.
+- A test driver component that can configure the Spin and WASI virtualization components, make one or more requests to the router, and then make assertions on the response returned and the state of the Spin and WASI virtualizations.
+
+The composition only has the following imports: 
+
+- A way to receive a `spin.toml` manifest file. This is needed by the router, Spin virtualization, and WASI virtualization components to know how they should be configured.
+- A few imports for working around limitations of guest support for `wasi:http`:
+    * An `http-helper` interface which allows creation of `wasi:http/types@0.2.0.{incoming-request, response-outparam, incoming-response}` resources since these can only be created by a host.
+    * Another function that can turn `wasi:http@types@0.2.0.{outgoing-response}` into a `wasi:http@types@0.2.0.{future-incoming-response}`
+
+*Note*: 
+
+> **A few notes:** 
+* We may still wish to have the composition import `wasi:cli/stdout` and `wasi:cli/stderr` so that it can easily log things.
```

Comment:
The example test I ran from the implementation repo did log to stdout -- is this already supported?